test/api/fake_api_test.dart

```diff
@@ -2,10 +2,13 @@ import 'package:checks/checks.dart';
 import 'package:test/scaffolding.dart';
 import 'package:zulip/api/exception.dart';
 
+import '../model/binding.dart';
 import 'exception_checks.dart';
 import 'fake_api.dart';
 
 void main() {
+  TestZulipBinding.ensureInitialized();
```

Comment:
@gnprice Thoughts on testing setup for this?

Because `ApiConnection` depends on `userAgent` function, and it depends on `ZulipBinding` to get package & device info — now every test that uses `ApiConnection` will need to also first call `TestZulipBinding.ensureInitialized()`.

Alternative implementation is to make `ApiConnection` take `userAgent` in it's constructor, which would just move this reposnibility to the upper layer (`LiveGlobalStore` for the real impl), but that would remove the need for `TestZulipBinding.ensureInitialized()` in each test.

Reply:
Hmm yeah, having to do that in all the tests would be annoying.

It looks like there's only a small number of call sites of the ApiConnection constructors. So that's probably a good solution.

Reply:
Hmm after trying the alternative I suggested, ended up reverting it, because it led to a bit of constructor argument poisoning up the chain of `ApiConnection` — making it harder to read the code where some constructors calls take the userAgent value, and some leave it out (null).

Instead ended up refactoring first implementation by moving the `userAgent` function to `class ZulipBinding` to make it explicit that it now depends on members in `ZulipBinding` and thus the binding needs to be initialized to get user-agent.

Reply:
What were the callers that that propagated up the call chain to? I was imagining it'd be just that `ApiConnection.live` requires a real user-agent, and `FakeApiConnection` defaults to a boring fake one, and then very little other code has to get involved.

Reply:
The implementation I tried was to keep `deviceInfo` and `packageInfo` as async function instead of the current sync getters on `ZulipBinding` and then trying to populate the ApiConnection args (deviceInfo & packageInfo) from the async functions it's being called. But there was [one call](https://github.com/zulip/zulip-flutter/blob/99f6a8b944248f01009e1959ff0f64e5c48789e5/lib/model/store.dart#L225) which was not async, so had to add the args for it too. And doing so led to many other changes in the codebase which were moot, thus reverted it.

But if we stick to sync function `userAgentHeader` and let `ApiConnection.live` inline it in it's constructor like:
```dart
ApiConnection.live({
  ...
}) : this(client: http.Client(),
          ...
          userAgentHeader: ZulipBinding.instance.userAgentHeader());
```
then it does work without the need to poison the callers, I'll send a revision with this now.

Reply:
Okay just tried it, and seems like there isn't much difference in line count, because now some tests that needs to check userAgent needs to pass the `testBinding.userAgent` to `FakeApiConnection.with_` which are quite a few. 

But on the plus side, the tests that don't check `userAgent` don't need to add ZulipBinding initialization lines:
```diff
-  TestZulipBinding.ensureInitialized();
-  tearDown(testBinding.reset);
```

Reply:
> Okay just tried it, and seems like there isn't much difference in line count, because now some tests that needs to check userAgent needs to pass the `testBinding.userAgent` to `FakeApiConnection.with_` which are quite a few.

Cool. Yeah, even if the total line count is the same but the impact is all concentrated in the one group of tests that's about the user-agent value itself, that sounds a lot nicer.

We can probably also improve that further with a local helper or two in those tests.